### ISSUE TITLE
[Feature Request] Filter branches to be deleted by regex (=inclusion list pattern)

### ISSUE THREAD

**Issue by jonringer** [hartwork/git-delete-merged-branches#52]
It would be great to filter the branches to be deleted by a regex.

In nixpkgs, we have a lot of generated `backport-....` branches which manifest in a PR. Eventually these PRs get merged, but often people forget to follow up and delete the branch. This creates a lot of dangling branches which exist in the upstream repo. It would be great for me to pass `--branch-regex 'backport-.*'` to select all of these.

**Comment by hartwork** [comment]
Hi @jonringer to be sure I understand: Would those `backport-.*` branches be branches to (a) always delete, (b) always keep, (c) be the only branches to inspect, (d) …? I'm not sure I understand the idea yet.

**Comment by hartwork** [comment]
@jonringer thanks for your reply. I'm still not sure I understand. `git-delete-merged-branches` would find those branches anyway if merged, so then the point would be to exclude all branches that do _not_ match the naming scheme and to save runtime on those?

**Comment by jonringer** [comment]
It's to mitigate risk of deleting branches which others may care. Essentially not to step on others toes

**Comment by hartwork** [comment]
@jonringer we currently have `--exclude <branch>` and what you're asking for then is `--include <branch>` but with regex powers added. Let me think about that; with UX in mind and all it's not as trivial as it may sound.

Note to self: `git branch -a '--format=%(refname:lstrip=2)' --list \*/backport-\*`

**Comment by jonringer** [comment]
Yea, my initial thought process would be to feed in the pattern to pythons regex, and use the matches as a way to filter candidates

**Comment by hartwork** [comment]
I would like to document possible workarounds:

- a) Mass-exclude all other branches (i.e. flipping inclusion around), e.g. using `git branch` output and a shell loop calling `git config "branch.${branch_name}.dmb-excluded" true` for all non-backport branches. You could run `git-delete-merged-branches --configure` afterwards to validate in the UI that the exclusion list looks sane.
- b) Run without `--yes|-y` and confirm/filter deletion of branches interactively.
- c) Check back with the team if deletion of all branches merged into X, Y, Z is considered acceptable.

**Comment by jonringer** [comment]
On this scale it's kind of hard to do any manual filtering. Here's some numbers:

```
[16:55:15] jon@jon-desktop /home/jon/projects/nixpkgs (spire-module)
$ git branch --list --all | grep origin | wc -l
259
[16:55:20] jon@jon-desktop /home/jon/projects/nixpkgs (spire-module)
$ git branch --list --all | grep origin | grep backport | wc -l
133
```
so there's close to 130 branches that I would need to manually exclude. New branches will come and go, so I can't make any assumptions about what branches will be there between runs.

Some merged branches also need to stick around, as there's additional infrastructure which is expecting the branches to exist, and constantly polling to check for updates.

**Comment by hartwork** [comment]
> On this scale it's kind of hard to do any manual filtering.

I'm aware, I suggested an *automated* approach using a shell loop. I can demo the approach if it's not clear. Marking all branches as excluded that do not match the pattern is easy using Bash and the `git config [..]` call mentioned above.

> ```
> [16:55:15] jon@jon-desktop /home/jon/projects/nixpkgs (spire-module)
> $ git branch --list --all | grep origin | wc -l
> 259
> [16:55:20] jon@jon-desktop /home/jon/projects/nixpkgs (spire-module)
> $ git branch --list --all | grep origin | grep backport | wc -l
> 133
> ```

It's big, I know, yes.

> 
> so there's close to 130 branches that I would need to manually exclude. New branches will come and go, so I can't make any assumptions about what branches will be there between runs.

Bash automaaaaaaation :smiley: 

> Some merged branches also need to stick around, as there's additional infrastructure which is expecting the branches to exist, and constantly polling to check for updates.

The exclude feature already handles that.

**Comment by jonringer** [comment]
For my use case of conservative pruning, this works quite well:
```
git remote prune origin; git branch -r --merged origin/release-21.11 | grep origin | grep backport | sed -e 's|/| |' | sed -e 's|  ||' | xargs  -P 10 -l bash -c 'git push -d $0 $1'
```

**Comment by hartwork** [comment]
Hi @jonringer ,

There is a pull request #77 now that adds a new command line argument `--include-regex=PATTERN` to implement a regex based inclusion list. Please be invited for testing and/or code review. `pip install -e .` from a Git clone should do, after switching to branch `improve-exclusion-inclusion`.

> For my use case of conservative pruning, this works quite well:
> 
> ```
> git remote prune origin; git branch -r --merged origin/release-21.11 | grep origin | grep backport | sed -e 's|/| |' | sed -e 's|  ||' | xargs  -P 10 -l bash -c 'git push -d $0 $1'
> ```

I believe it's a different level of risk to mass-run `git push -d` from shell than to auto-create an exclusion list, hence my personal recommendation against former and pro latter, for safety. I felt that needed saying.

PS: To add a concrete example to https://github.com/hartwork/git-delete-merged-branches/issues/52#issuecomment-1037500921 for anyone intererested: To mass-exclude certain branches based on a regex pattern in a Bash world prior to (but also after) pull request #77:

```bash
for i in $(git branch --format='%(refname:lstrip=2)'); do
    [[ ${i} =~ ^some-regex-here$ ]] || continue
    git config "branch.${i}.dmb-excluded" true
done
```

Best, Sebastian

**Comment by hartwork** [comment]
@jonringer any chance you could have a look at pull request #77 this week?
